src/apply.c

```diff
@@ -480,3 +483,110 @@ int git_apply_tree(
 
 	return error;
 }
+
+/* normal: apply to workdir: ignore index
+ * --cached: apply to index: ignore workdir
+ * --index: apply to both: validate index == workdir
+ */
+
+int git_apply(
+	git_repository *repo,
+	git_diff *diff,
+	git_apply_options *given_opts)
+{
+	git_index *contents = NULL, *repo_index = NULL;
+	git_reader *pre_reader = NULL;
+	const git_diff_delta *delta;
+	git_vector paths = GIT_VECTOR_INIT;
+	git_apply_options opts = GIT_APPLY_OPTIONS_INIT;
+	git_checkout_options checkout_opts = GIT_CHECKOUT_OPTIONS_INIT;
+	bool do_checkout;
+	size_t i;
+	int error;
+
+	assert(repo && diff);
+
+	GITERR_CHECK_VERSION(
+		given_opts, GIT_APPLY_OPTIONS_VERSION, "git_apply_options");
+
+	if (given_opts)
+		memcpy(&opts, given_opts, sizeof(git_apply_options));
```

Comment:
(FWIW I remember doing it the other way around : `if (given_opts == NULL) *given_opts = opts;`. Saves a memcpy 🙄)

Reply:
The nice thing about creating a copy is that you can [set or override options](https://github.com/libgit2/libgit2/blob/master/src/merge.c#L2066) [that the caller didn't provide](https://github.com/libgit2/libgit2/blob/e9b2e4343f636d0cdb030ffe6ec1edc76d344253/src/diff_tform.c#L246).  We don't do that here yet, but I suspect that we will want to over time.